NOINLINE some pid.c functions to reduce ITCM usage

ctzsnooze opened the pull request:
Intended to reduce ITCM RAM usage, which is very constrained in H7 and F745 processors since updating to ARM compiler 13.3 rel1.

I don't know how else to solve the problem, but my PR #13975 suddenly won't fit on these processors, whereas it did until the 2-3 days.

haslinghuis commented:
Without 53d44aa1b15c609dc704bad654f7f2dc3ec7058a it build Poshold PR but ARM SDK update (just merged) might add up.

ctzsnooze commented:
Unfortunately, when I added this to PR #13975, the ITCM RAM usage went up from 101.71% to 101.73%, so whatever I did here is achieving nothing useful.

Hopefully others here know how to solve this problem most effectively, because I don't have a clue :-)

blckmn reviewed (changes requested):
Use the tag `FAST_CODE_PREF` and that will remove this from ITCM for those with limited ITCM

ctzsnooze commented:
@blckmn - very strange errors when trying to apply `FAST_CODE_PREF`; am having to revert most of these...

compiling ../main/flight/pid.c 
../main/flight/pid.c:788:20: error: unknown type name 'FAST_CODE_PREF'
STATIC_UNIT_TESTED FAST_CODE_PREF void rotateItermAndAxisError(void)
                   ^
../main/flight/pid.c:820:20: error: unknown type name 'FAST_CODE_PREF'
STATIC_UNIT_TESTED FAST_CODE_PREF void applyAbsoluteControl(const int axis, const float gyroRate, float *currentPidSetpoint, float *itermErrorRate)
                   ^
../main/flight/pid.c:1024:28: error: expected ';' after top level declarator
static float FAST_CODE_PREF getSterm(int axis, const pidProfile_t *pidProfile, float setpoint)
                           ^
                           ;
3 errors generated.

blckmn commented:
> @blckmn - very strange errors when trying to apply `FAST_CODE_PREF`; am having to revert most of these...
> 
> ```
> compiling ../main/flight/pid.c 
> ../main/flight/pid.c:788:20: error: unknown type name 'FAST_CODE_PREF'
> STATIC_UNIT_TESTED FAST_CODE_PREF void rotateItermAndAxisError(void)
>                    ^
> ../main/flight/pid.c:820:20: error: unknown type name 'FAST_CODE_PREF'
> STATIC_UNIT_TESTED FAST_CODE_PREF void applyAbsoluteControl(const int axis, const float gyroRate, float *currentPidSetpoint, float *itermErrorRate)
>                    ^
> ../main/flight/pid.c:1024:28: error: expected ';' after top level declarator
> static float FAST_CODE_PREF getSterm(int axis, const pidProfile_t *pidProfile, float setpoint)
>                            ^
>                            ;
> 3 errors generated.
> ```
> 
> and these
> 
> ```
> compiling ../main/flight/pid.c 
> ../main/flight/pid.c:788:20: error: unknown type name 'FAST_CODE_PREF'
> STATIC_UNIT_TESTED FAST_CODE_PREF void rotateItermAndAxisError(void)
>                    ^
> ../main/flight/pid.c:1043:22: error: variable has incomplete type 'void'
> NOINLINE static void FAST_CODE_PREF calculateSpaValues(const pidProfile_t *pidProfile)
>                      ^
> ../main/flight/pid.c:1043:36: error: expected ';' after top level declarator
> NOINLINE static void FAST_CODE_PREF calculateSpaValues(const pidProfile_t *pidProfile)
>                                    ^
>                                    ;
> 3 errors generated.
> ```

This is building your PR 13975? or this PR? Not clear here.

blckmn commented:
Add a define for FAST_CODE_PREF in platform.h in the unit test.

limonspb commented:
I thought FAST_CODE_PREF makes functions to be more likely placed in ITCM_RAM thus increases it's usage?
When i build locally with and without this PR with `make hex REV=yes TARGET=STM32F745` it gives same result:
```
ITCM_RAM:       15884 B        16 KB     96.95%
```

ctzsnooze commented:
Squashed this for clarity. 
Applied `FAST_CODE_PREF` to the wing functions and pidaudio, added some comments.
Rebased to master, updated the toolchain, checked that it built OK for F411 etc as usual.

Checked that  F745 target.h  has `#define FAST_CODE_PREF`  so that functions tagged `FAST_CODE_PREF` would not go to ITCM RAM for this target.  Then built using `make STM32F745 EXTRA_FLAGS="-D'RELEASE_NAME=4.6.0-dev' -DCORE_BUILD"`Result:

PR 14048:
```
 Linking STM32F745 
Memory region         Used Size  Region Size  %age Used
        ITCM_RAM:         15 KB        16 KB     93.75%
      ITCM_FLASH:         456 B        32 KB      1.39%
ITCM_FLASH_CONFIG:           0 B        32 KB      0.00%
     ITCM_FLASH1:      550552 B       960 KB     56.01%
      AXIM_FLASH:         456 B        32 KB      1.39%
AXIM_FLASH_CONFIG:           0 B        32 KB      0.00%
     AXIM_FLASH1:      577819 B       960 KB     58.78%
        DTCM_RAM:       43788 B        64 KB     66.82%
           SRAM1:       85220 B       240 KB     34.68%
           SRAM2:           0 B        16 KB      0.00%
       MEMORY_B1:           0 B          0 B
   text       data        bss        dec        hex    filename
 569627       8648     120356     698631      aa907    ./obj/main/betaflight_STM32F745.elf
```

 Master
 
```
Linking STM32F745 
Memory region         Used Size  Region Size  %age Used
        ITCM_RAM:         15 KB        16 KB     93.75%
      ITCM_FLASH:         456 B        32 KB      1.39%
ITCM_FLASH_CONFIG:           0 B        32 KB      0.00%
     ITCM_FLASH1:      550552 B       960 KB     56.01%
      AXIM_FLASH:         456 B        32 KB      1.39%
AXIM_FLASH_CONFIG:           0 B        32 KB      0.00%
     AXIM_FLASH1:      577819 B       960 KB     58.78%
        DTCM_RAM:       43788 B        64 KB     66.82%
           SRAM1:       85220 B       240 KB     34.68%
           SRAM2:           0 B        16 KB      0.00%
       MEMORY_B1:           0 B          0 B
   text       data        bss        dec        hex    filename
 569627       8648     120356     698631      aa907    ./obj/main/betaflight_STM32F745.elf
```

@blckmn - I don't understand why there's no change in ITCM Ram usage?

ledvinap commented:
FAST_CODE_PREF /  FAST_CODE does different thing - it says that function should be placed in `.tcm_code` section, possibly preferably.
So adding FAST_CODE_PREF to function on target that that does not define is is NOP and it forces optimization on targets with enough memory.
So this PR changes nothing.

Functions are placed in TCM RAM because they are inlined into function that is defined as FAST_CODE.
ITCM_RAM_OTIMISATION tries to reduce memory pressure by decreasing optimization level to -O2 (which will also disable optimization that make code smaller). I can't find any reference about gcc -flto behavior when inlining function - does it inherit caller optimization settings?

FAST_CODE_NOINLINE is actually macro that **disables** FAST_CODE optimization - it does force noinline function, so code is not in tcm_code and file-level optimization setting _may_ apply. But there is (possibly considerable) cost associated with `noiniline` that does apply even when function is not called from FAST_CODE.

ledvinap commented:
IMO these are functions (symbols) that take up space in TCM_RAM:
```address size symbol
00000000 000000c0 T pwmCompleteDshotMotorUpdate
00000000 T tcm_code_start
000000c0 000001d8 t decode_bb
00000298 000002c8 T motor_DMA_IRQHandler.lto_priv.0
00000560 000000d0 T pwmWriteDshotInt
00000630 000002bc T spiSequenceStart
00000800 A _Min_Stack_Size
000008ec 0000001a T pt1FilterApply
00000908 0000002a T pt2FilterApply
00000934 0000003a T pt3FilterApply
00000970 0000004e T biquadFilterApplyDF1
000009c0 00000042 T biquadFilterApply
00000a04 00000008 t run
00000a0c 00000010 T pt2FilterGain
00000a1c 00000010 T pt3FilterGain
00000a2c 00000052 T laggedMovingAverageUpdate
00000a80 000000a4 T biquadFilterUpdate
00000b24 0000008c T biquadFilterUpdateLPF
00000bb0 00000004 T dshotWriteInt.lto_priv.0
00000bb4 00000016 T dshotWrite.lto_priv.0
00000bcc 00000040 T prepareDshotPacket
00000c0c 00000020 T dshotCommandQueueFull.lto_priv.0
00000c2c 0000001c T dshotCommandQueueEmpty
00000c48 00000020 T isLastDshotCommand.lto_priv.0
00000c68 00000044 T dshotCommandIsProcessing
00000cac 00000048 T dshotCommandQueueUpdate.lto_priv.0
00000cf4 00000014 T dshotCommandCyclesFromTime.lto_priv.0
00000d08 00000278 t schedulerExecuteTask
00000f80 0000017c T taskGyroSample
000010fc 00000934 T taskFiltering
00001a30 0000000c T schedulerSetNextStateTime
00001a3c 00000010 T schedulerGetNextStateTime
00001a4c 00000b20 T scheduler
0000256c 00000780 T taskMainPidLoop
00002cec 0000050c T processRcCommand
000031f8 00000aa8 T pidController
00003ca0 00000130 T gyroUpdateSensor.lto_priv.0
00003dd0 00000128 T sdftWinSq.constprop.0
00003ef8 T tcm_code_end

ledvinap commented:
From gcc docs (https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html) : The optimize attribute should be used for debugging purposes only. It is not suitable in production code.

IIRC the same applies to mixing different optimization levels using LTO ..

limonspb commented:
@ledvinap could you please share which command did you use to see functions in TCM_RAM with sizes?

ledvinap commented:
` arm-none-eabi-nm -S --numeric-sort obj/main/betaflight_STM32F745.elf` and decode section based on address

ctzsnooze commented:
@ledvinap - thanks for helping clarify things here.

My understanding of `FAST_CODE_PREF` has improved a little in recent hours, hopefully.

Could you possibly clarify if I understand it correctly?

Currently, the F745 and F7X2 target files include a line `#define FAST_CODE_PREF`.  My understanding is the *intent* of this is that, *for these targets only*, a function decorated with `FAST_CODE_PREF` will *not* go into ITCM RAM; however, in contrast, a target which lacks the `#define FAST_CODE_PREF` line, *will* move the function into ITCM RAM.

I just repeated my earlier test.  The build string includes wing, pos_hold, etc etc a fairly typical build:
```
make JHEF745 EXTRA_FLAGS="-D'RELEASE_NAME=4.6.0-zulu' -DCLOUD_BUILD -DUSE_ALT_HOLD_MODE -DUSE_DSHOT -DUSE_GPS -DUSE_GPS_PLUS_CODES -DUSE_MAG -DUSE_OSD -DUSE_OSD_SD -DUSE_POS_HOLD_MODE -DUSE_WING -DUSE_SERIALRX -DUSE_SERIALRX_CRSF -DUSE_TELEMETRY -DUSE_TELEMETRY_CRSF -DUSE_VTX"
```

This is allocation in master
```
Linking STM32F745_JHEF745 
Memory region         Used Size  Region Size  %age Used
        ITCM_RAM:       15068 B        16 KB     91.97%
      ITCM_FLASH:         456 B        32 KB      1.39%
ITCM_FLASH_CONFIG:           0 B        32 KB      0.00%
     ITCM_FLASH1:      395472 B       960 KB     40.23%
      AXIM_FLASH:         456 B        32 KB      1.39%
AXIM_FLASH_CONFIG:           0 B        32 KB      0.00%
     AXIM_FLASH1:      419743 B       960 KB     42.70%
        DTCM_RAM:       39376 B        64 KB     60.08%
           SRAM1:       55124 B       240 KB     22.43%
           SRAM2:           0 B        16 KB      0.00%
       MEMORY_B1:           0 B          0 B
   text	   data	    bss	    dec	    hex	filename
 413603	   6596	  87900	 508099	  7c0c3	./obj/main/betaflight_STM32F745_JHEF745.elf
 ```
 
 vs with this PR
 ```
 Linking STM32F745_JHEF745 
Memory region         Used Size  Region Size  %age Used
        ITCM_RAM:       15068 B        16 KB     91.97%
      ITCM_FLASH:         456 B        32 KB      1.39%
ITCM_FLASH_CONFIG:           0 B        32 KB      0.00%
     ITCM_FLASH1:      395472 B       960 KB     40.23%
      AXIM_FLASH:         456 B        32 KB      1.39%
AXIM_FLASH_CONFIG:           0 B        32 KB      0.00%
     AXIM_FLASH1:      419743 B       960 KB     42.70%
        DTCM_RAM:       39376 B        64 KB     60.08%
           SRAM1:       55124 B       240 KB     22.43%
           SRAM2:           0 B        16 KB      0.00%
       MEMORY_B1:           0 B          0 B
   text	   data	    bss	    dec	    hex	filename
 413603	   6596	  87900	 508099	  7c0c3	./obj/main/betaflight_STM32F745_JHEF745.elf
```

It seems that despite applying `FAST_CODE_PREF` to all wing functions in pid.c, there is absolutely no change in ITCM RAM usage.

SteveCEvans commented:
> @ledvinap - thanks for helping clarify things here.
> 
> My understanding of `FAST_CODE_PREF` has improved a little in recent hours, hopefully.
> 
> Could you possibly clarify if I understand it correctly?
> 
> Currently, the F745 and F7X2 target files include a line `#define FAST_CODE_PREF`. My understanding is the _intent_ of this is that, _for these targets only_, a function decorated with `FAST_CODE_PREF` will _not_ go into ITCM RAM; however, in contrast, a target which lacks the `#define FAST_CODE_PREF` line, _will_ move the function into ITCM RAM.
> 
> I just repeated my earlier test. The build string includes wing, pos_hold, etc etc a fairly typical build:
> 
> ```
> make JHEF745 EXTRA_FLAGS="-D'RELEASE_NAME=4.6.0-zulu' -DCLOUD_BUILD -DUSE_ALT_HOLD_MODE -DUSE_DSHOT -DUSE_GPS -DUSE_GPS_PLUS_CODES -DUSE_MAG -DUSE_OSD -DUSE_OSD_SD -DUSE_POS_HOLD_MODE -DUSE_WING -DUSE_SERIALRX -DUSE_SERIALRX_CRSF -DUSE_TELEMETRY -DUSE_TELEMETRY_CRSF -DUSE_VTX"
> ```
> 
> This is allocation in master
> 
> ```
> Linking STM32F745_JHEF745 
> Memory region         Used Size  Region Size  %age Used
>         ITCM_RAM:       15068 B        16 KB     91.97%
>       ITCM_FLASH:         456 B        32 KB      1.39%
> ITCM_FLASH_CONFIG:           0 B        32 KB      0.00%
>      ITCM_FLASH1:      395472 B       960 KB     40.23%
>       AXIM_FLASH:         456 B        32 KB      1.39%
> AXIM_FLASH_CONFIG:           0 B        32 KB      0.00%
>      AXIM_FLASH1:      419743 B       960 KB     42.70%
>         DTCM_RAM:       39376 B        64 KB     60.08%
>            SRAM1:       55124 B       240 KB     22.43%
>            SRAM2:           0 B        16 KB      0.00%
>        MEMORY_B1:           0 B          0 B
>    text	   data	    bss	    dec	    hex	filename
>  413603	   6596	  87900	 508099	  7c0c3	./obj/main/betaflight_STM32F745_JHEF745.elf
> ```
> 
> vs with this PR
> 
> ```
> Linking STM32F745_JHEF745 
> Memory region         Used Size  Region Size  %age Used
>        ITCM_RAM:       15068 B        16 KB     91.97%
>      ITCM_FLASH:         456 B        32 KB      1.39%
> ITCM_FLASH_CONFIG:           0 B        32 KB      0.00%
>     ITCM_FLASH1:      395472 B       960 KB     40.23%
>      AXIM_FLASH:         456 B        32 KB      1.39%
> AXIM_FLASH_CONFIG:           0 B        32 KB      0.00%
>     AXIM_FLASH1:      419743 B       960 KB     42.70%
>        DTCM_RAM:       39376 B        64 KB     60.08%
>           SRAM1:       55124 B       240 KB     22.43%
>           SRAM2:           0 B        16 KB      0.00%
>       MEMORY_B1:           0 B          0 B
>   text	   data	    bss	    dec	    hex	filename
> 413603	   6596	  87900	 508099	  7c0c3	./obj/main/betaflight_STM32F745_JHEF745.elf
> ```
> 
> It seems that despite applying `FAST_CODE_PREF` to all wing functions in pid.c, there is absolutely no change in ITCM RAM usage.

Yeah, that's what I just explained in chat...

ctzsnooze commented:
@SteveCEvans - given that understanding of how `FAST_CODE_PREF` works, should I have expected a change in ITCM RAM with this PR when the build includes WING, and all the wing functions have the `FAST_CODE_PREF` label?  That doesn't seem to happen when making the build locally.  

Could my failure here be because perhaps my local build method didn't explicitly set the compiler optimisations?  

Should local builds from master utilise the same optimisations as the online build?  That's something I guess I need to know.

SteveCEvans commented:
You can do a "cloud build" locally using `make <target> EXTRA_FLAGS="-DCLOUD_BUILD"`.

SteveCEvans commented:
Do both builds and then compare the `obj/main/betaflight_STM32F745_<target>_AIO.map` files, looking at the section which starts thus:

```
.tcm_code       0x00000000     0x2dc4 load address 0x08049b80
                0x00000000                        . = ALIGN (0x4)
                0x00000000                        tcm_code_start = .
 *(.tcm_code)
 .tcm_code      0x00000000      0x608 /var/folders/7k/11ftt0b56792cylqxw59w69m0000gp/T//ccmh0y95.ltrans0.ltrans.o
                0x00000000                pt1FilterApply
                0x0000001c                pt2FilterApply
                0x00000048                pt3FilterApply
                0x00000084                biquadFilterApplyDF1
                0x000000d4                biquadFilterApply
```

Then you can see which functions have been added to `ITCM`.

ctzsnooze commented:
Thanks @SteveCEvans I've just done that exactly, using cloud build for F745 and including Wing, and there is zero difference in the .tcm assignment, function by function, when comparing master to the changes in this PR.  

On the other hand, when comparing master to the pos_hold PR, there are changes - big changes - in .tcm function allocations affecting nearly every .tcm function - Dshot, Scheduler, you name it, they all change.  I'll post those results in that PR and maybe we can work out why such a PR, which should, on the face of it, have little or nothing to do with ITCM RAM, actually makes such a difference.

ctzsnooze commented:
closing since root cause of the problem was found - see PR #13975
